### DISCUSSION TITLE
Where are the thumbnails?

### DISCUSSION THREAD

**Discussion by Bendermk6** [berrystephenw/OctoText#70] [General]
Hey,
i just got the 0.2.0 release installed and wonder how to get thumbnails in my notification mails?

I see in the logfile that octotext is searching for thumbnails in .gcode files from prusaslicer and .ufp files from cura, but i don't get any thumbnail in the notification? I have "Cura Thumbnails" and "PrusaSlicer Thumbnails" installed.

**Comment by berrystephenw** [comment]
Hi, to get thumbnails in your notifications OctoText will only know about them when you upload new files. It doesn't check for existing plugins and use thumbnails if they already exist. My understanding of how those plugins work is to strip the thumbnails from the gcode files and store them locally in the plugin's local directory structure - it would have been a much larger task for me to check for existing plugins and look in their directory structure for a matching thumbnail. Instead on every upload the file is checked and a thumbnail if it exists is copied into Octotext's local directory and used. 
I could have also scanned the existing files that have been uploaded and copied them - but I was very worried about the performance impact where people have hundreds of gcode files and searching for embedded thumbnails. 

I hope this answers your question, I am assuming the file wasn't newly uploaded after you installed OctoText. If you did upload the file and you know it has a thumbnail embedded, please attach it here and I'll see what is going on.

**Comment by Bendermk6** [comment]
Hey,
okay, that makes completely sense that you don't need other plugins to get your thumbnails.
I figured out that i should read your readme.md -> _The thumbnail will only be sent on a print START event._

So thumbnails in a .gcode file generated from PrusaSlicer 2.3.0 are working fine -> with print start a got it in the email.

I also tried to get some from the .ufp files from Cura 4.9, without success.
The uploaded .ufp file has a .png inside (opened with 7zip on win10) and are shown in Octoprint with the CuraThumbnails plugin without problems.
But as i saw in the logfile, OctoText seems to get something wrong with the path to the saved .png file.

Here my log files with some debugging on and the Cura .ufp file.
![Screenshot 2021-04-23 195359](https://user-images.githubusercontent.com/81684925/115910905-b3ced800-a46d-11eb-9444-ed445488f1be.png)
[octoprint.log](https://github.com/berrystephenw/OctoText/files/6367218/octoprint.log)


And for the performance problems you mentioned, i uploaded yesterday a ~90mb MMU-Print .gcode file and my Pi4B 4GB got a 100% CPU usage from gcode analysis, unfortunately i deleted the log. I only have a screenshot from htop:
![octo2](https://user-images.githubusercontent.com/81684925/115911260-293aa880-a46e-11eb-8913-f4a546cae7f7.png)

I don't know if that was the searching for the thumbnail from OctoText or some other plugin. I can retry the ~90mb upload tomorrow again with some proper logfile.

Thanks again for OctoText :)

**Reply by berrystephenw** [reply]
Thanks I appreciate the log files, they are helpful! So first, the performance problem you are noticing is not OctoText ;) I believe that is the analysis done on the file to figure out how much time it takes to print the model. The bigger the file, the more time it takes to analyze.

The log file shows this:
`2021-04-23 18:30:03,558 - octoprint.plugins.OctoText - DEBUG - Line 25: Detected first extrusion. Read complete.
2021-04-23 18:30:03,559 - octoprint.plugins.OctoText - DEBUG - ;FLAVOR:Marlin
;TIME:3667
;Filament used: 6.40537m
;Layer height: 0.25
;MINX:73.112
;MINY:65.375
;MINZ:0.25
;MAXX:176.887
;MAXY:144.625
;MAXZ:9
;LINEARADVANCEPROCESSED
;Generated with Cura_SteamEngine 4.9.0

2021-04-23 18:30:03,561 - octoprint.plugins.OctoText - DEBUG - thumbnail filename /home/pi/.octoprint/data/OctoText/PI3MK2_trenner_blende_vorne.png`

Normally you would see what looks like dozens of strings of garbage text - which is the icon embedded in the file. Since that wasn't output in the log, the code didn't find what it was looking for and probably generated an empty png file. You could if you were curious go to /home/pi/.octoprint/data/OctoText/PI3MK2_trenner_blende_vorne.png and see if there was anything in the file. 

I'll download Cura and see if I can generate a test file. Probably should have done this before!

**Reply by berrystephenw** [reply]
Ok, I have a test file and it behaves the same. I'll work on it an publish an update. Thanks for testing! @Bendermk6

**Reply by Bendermk6** [reply]
Oh crap, just noticed I forgot to attach the .ufp file.

But yes I saw octotext tries to searched for the thumbnail like in the prusa flies.

**Reply by berrystephenw** [reply]
Ok, I talked to the developer of the ultimaker plug-in and he gave me some hints. First, he coordinated with the Cura octoprint plug-in developer and the two of them work together. What this means is that a direct network load from cura to the ultimaker plug-in only works for that combination, but I can however take the manually uploaded ULP files and process them.
There was a bug with the upload of ULP file types that I've fixed, I'll release a new rc build this weekend.
More than likely I'll change how the plug-in does thumb nails and require the ultimaker/prusa plugins and take the existing thumbs from  there...
![image](https://user-images.githubusercontent.com/44931130/115944665-0d8ecc80-a485-11eb-86f5-c505c7edd30f.png)
